test/transforms.js

```diff
@@ -0,0 +1,156 @@
+  const diffs = await toArray(m2)
+
+  t.ok(diffs.length > 0, 'emitted some diffs')
+  for (const d of diffs) t.is(d.op, 'change')
+})
+
+function errorTransform () {
+  let done = false
+  return new Transform({
+    transform (chunk, cb) {
+      if (!done) {
+        done = true
+        cb(new Error('transform boom'))
+      } else {
+        this.push(chunk)
+        cb(null)
+      }
+    }
+  })
+}
+
+test('transforms: errors propagate and abort mirror', async function (t) {
+  const { local: src } = await createDrives(t)
+  const { local: dst } = await createDrives(t, { setup: false })
+
+  await src.put('/notes.txt', b4a.from('hello'))
+
+  const transformers = [(key) => errorTransform()]
+
+  const m = new MirrorDrive(src, dst, { transformers })
+
+  try {
+    await m.done()
+    t.fail('should have thrown from transform')
+  } catch (err) {
+    t.is(err.message, 'transform boom')
+  }
+})
+
+test('transformers: passthrough (null) still writes on rerun', async function (t) {
+  const { local: src } = await createDrives(t)
+  const { local: dst } = await createDrives(t, { setup: false })
+
+  const passthrough = (key) => null
+
+  const m1 = new MirrorDrive(src, dst, { transformers: [passthrough] })
+  await m1.done()
+
+  const m2 = new MirrorDrive(src, dst, { transformers: [passthrough], includeEquals: true })
+  const diffs = await toArray(m2)
+
+  t.ok(diffs.length > 0, 'emitted some diffs')
+  for (const d of diffs) t.is(d.op, 'change')
+})
+
+test('transformers: length-changing transform writes on rerun', async function (t) {
+  const { local: src } = await createDrives(t)
+  const { local: dst } = await createDrives(t, { setup: false })
+
+  // Increase length by appending fruit to every chunk
+  function fruitSaladTransform () {
+    return new Transform({
+      transform (chunk, cb) {
+        this.push(b4a.concat([chunk, b4a.from('🍐')]))
+        cb(null)
+      }
+    })
+  }
+
+  const transformers = [(key) => fruitSaladTransform()]
+
+  const m1 = new MirrorDrive(src, dst, { transformers })
+  await m1.done()
+
+  const m2 = new MirrorDrive(src, dst, { transformers, includeEquals: true })
+  const diffs = await toArray(m2)
+
+  t.ok(diffs.length > 0, 'emitted some diffs')
+  for (const d of diffs) t.is(d.op, 'change')
+})
+
+test('dry run + transforms: emits adds but writes nothing', async function (t) {
+  const { local: src } = await createDrives(t)
+  const { local: dst } = await createDrives(t, undefined, { setup: false })
+
+  const transformers = [(key) => upperTransform()]
+
+  const m = new MirrorDrive(src, dst, { dryRun: true, transformers, includeEquals: true })
+  const diffs = await toArray(m)
+
+  t.ok(diffs.length > 0, 'emitted some diffs')
+  for (const d of diffs) t.is(d.op, 'add')
+
+  t.absent(await dst.entry('/equal.txt'))
+})
+
+test('transforms + symlink: creates symlink and rerun yields change', async function (t) {
```

Comment:
Going with same strategy, ie if transforms then we can't be sure without running so assume change.